backends/apple/mps/runtime/MPSExecutor.mm

```diff
@@ -157,43 +147,41 @@ @interface MPSGraphExecutable()
 MPSExecutor::updateDataBuffers(
   std::vector<const Tensor*>& inputs, std::vector<const Tensor*>& outputs
 ) {
-  if (!_buffers_initialized) {
-    for (int i = 0; i < inputs.size(); i++) {
-      const Tensor& tensor = *inputs[i];
-      void* host_src = tensor.mutable_data_ptr<void*>();
-      if (_use_shared_mem) {
-        // Use directly the CPU buffer when using shared memory.
-        _inputGPUBuffers[i] = getMTLBufferStorage(tensor);
-      } else {
-        _inputCPUBuffers[i].flags = 0;
+  for (int i = 0; i < inputs.size(); i++) {
+    const Tensor& tensor = *inputs[i];
+    void* host_src = tensor.mutable_data_ptr<void*>();
+    if (_use_shared_mem) {
+      // Use directly the CPU buffer when using shared memory.
+      _inputGPUBuffers[i] = getMTLBufferStorage(tensor);
+    } else {
+      _inputCPUBuffers[i].flags = 0;
 #if TARGET_OS_SIMULATOR
-        // Simulator crashes when using newBufferWithBytesNoCopy.
-        // Use memcpy directly instead of using blit to copy the CPU
-        // data into the GPU buffer.
-        _inputCPUBuffers[i].srcOffset = 0;
-        _inputCPUBuffers[i].srcBuffer = host_src;
-        _inputCPUBuffers[i].srcCpu = 1;
+      // Simulator crashes when using newBufferWithBytesNoCopy.
+      // Use memcpy directly instead of using blit to copy the CPU
+      // data into the GPU buffer.
+      _inputCPUBuffers[i].srcOffset = 0;
+      _inputCPUBuffers[i].srcBuffer = host_src;
+      _inputCPUBuffers[i].srcCpu = 1;
 #else
-        MTLResourceOptions options = MTLResourceCPUCacheModeDefaultCache | MTLResourceStorageModeShared;
-        NSUInteger alignedLength = 0;
-        void* alignedPtr = pageAlignedBlockPtr(host_src, (NSUInteger)tensor.nbytes(), &alignedLength);
-        _inputCPUBuffers[i].srcOffset = uintptr_t(host_src) - uintptr_t(alignedPtr);
-        _inputCPUBuffers[i].srcBuffer = [MPSDevice::getInstance()->device() newBufferWithBytesNoCopy:alignedPtr
-                                                          length:alignedLength
-                                                        options:options
-                                                    deallocator:nil];
+      MTLResourceOptions options = MTLResourceCPUCacheModeDefaultCache | MTLResourceStorageModeShared;
+      NSUInteger alignedLength = 0;
+      void* alignedPtr = pageAlignedBlockPtr(host_src, (NSUInteger)tensor.nbytes(), &alignedLength);
+      _inputCPUBuffers[i].srcOffset = uintptr_t(host_src) - uintptr_t(alignedPtr);
+      _inputCPUBuffers[i].srcBuffer = [MPSDevice::getInstance()->device() newBufferWithBytesNoCopy:alignedPtr
```

Comment:
whts the difference between this one and the _use_shared_mem if this one does not do copy?

Reply:
Previous to macOS Sonoma/iOS 17 `newBuferWithBytesNoCopy` required the CPU buffer to be page aligned to wrap it around newBufferWithBytesNoCopy. Since Sonoma / iOS17 this requirement is not needed anymore. This function (`updateDataBuffer` has two paths, one for > macOS Sonoma / iOS17 and one for older):
```
Error
MPSExecutor::updateDataBuffers(
   // ..
    if (_use_shared_mem) { // _use_shared_mem = true if on shared memory and >= macOS Sonoma / iOS 17
      // Use directly the CPU buffer when using shared memory.
      _inputGPUBuffers[i] = getMTLBufferStorage(tensor);  // no copy
    }
    else {
         _inputCPUBuffers[i].srcOffset = uintptr_t(host_src) - uintptr_t(alignedPtr);
         _inputCPUBuffers[i].srcBuffer = [MPSDevice::getInstance()->device() newBufferWithBytesNoCopy:alignedPtr
                                                        length:alignedLength
                                                      options:options
                                                  deallocator:nil];
     }
     // ..
    if (!_use_shared_mem) { // blit from CPU buffer into GPU buffer if on < macOS Sonoma / iOS 17 or non-shared memory device
      MPSStream* mpsStream = getDefaultMPSStream();
      mpsStream->copy_and_sync(
        _inputCPUBuffers, /*non_blocking=*/true);
  }
```

There is the memory planning option during AOT where we could set the alignment to PAGE_SIZE (4096) if we're running before macOS Sonoma / iOS 17, so there will be no copy at all even for older OSes (all CPU buffers would be  already page aligned during runtime). I've played a little bit and got it running some time ago but haven't had a chance to create a PR.